helix-contract/contracts/mapping-token/v2/message-handle/DarwiniaSub2SubMessageHandle.sol

```diff
@@ -0,0 +1,160 @@
+        dstStorageKeyForLastDeliveredNonce = _dstStorageKeyForLastDeliveredNonce;
+    }
+
+    function derivedRemoteSender(bytes4 srcSubstrateChainId, address sender) public view returns(address) {
+        return SmartChainXLib.deriveSenderFromRemote(
+            srcSubstrateChainId,
+            sender
+        );
+    }
+
+    function sendMessage(
+        uint256 remoteReceiveGasLimit,
+        uint32  remoteSpecVersion,
+        uint64  remoteCallWeight,
+        address receiver,
+        bytes calldata callPayload) external onlyCaller payable returns(uint256) {
+        PalletEthereum.MessageTransactCall memory call = PalletEthereum.MessageTransactCall(
+            remoteMessageTransactCallIndex,
+            PalletEthereum.buildTransactionV2ForMessageTransact(
+                remoteReceiveGasLimit, // gas limit
+                remoteHelix,
+                remoteSmartChainId,
+                abi.encodeWithSelector(
+                    this.recvMessage.selector,
+                    receiver,
+                    callPayload 
+                )
+            )
+        );
+        bytes memory callEncoded = PalletEthereum.encodeMessageTransactCall(call);
+        bytes memory messagePayload = SmartChainXLib.buildMessage(
+            remoteSpecVersion,
+            remoteCallWeight,
+            callEncoded
+        );
+
+        SmartChainXLib.sendMessage(
+            dispatchAddress,
+            callIndexOfSendMessage,
+            outboundLaneId,
+            msg.value,
+            messagePayload
+        );
+
+        uint64 nonce = SmartChainXLib.latestNonce(
+            storageAddress,
+            srcStorageKeyForLatestNonce,
+            outboundLaneId
+        );
+
+        return encodeTransferId(outboundLaneId, nonce);
+    }
+
+    function recvMessage(address receiver, bytes calldata callPayload) external onlyRemoteHelix whenNotPaused {
+        require(hasRole(CALLER_ROLE, receiver), "DarwiniaSub2SubMessageHandle:receiver is not caller");
+        (bool result,) = receiver.call(callPayload);
+        require(result, "DarwiniaSub2SubMessageHandle:call app failed");
+    }
+
+    function encodeTransferId(bytes4 laneId, uint64 nonce) public pure returns(uint256) {
+        return (uint256(uint32(laneId)) << 64) + uint256(nonce);
+    }
+
+    function decodeTransferId(uint256 transferId) public pure returns(bytes4, uint64) {
+        return (bytes4(uint32(transferId >> 64)), uint64(transferId & 0xffffffffffffffff));
+    }
+
+    function isMessageTransfered(uint256 transferId) public view returns(bool) {
+        (bytes4 laneId, uint64 nonce) = decodeTransferId(transferId);
+        uint64 latestNonce = SmartChainXLib.lastDeliveredNonce(
+            storageAddress,
+            dstStorageKeyForLastDeliveredNonce,
+            laneId
+        );
+        return nonce <= latestNonce;
+    }
+
+    function latestRecvMessageId() public view returns(uint256) {
+        return SmartChainXLib.lastDeliveredNonce(
```

Comment:
The returned `lastDeliveredNonce ` is `laneId + nonce`, then the naming is a problem ?